diff --git a/src/JMS/Serializer/JsonSerializationVisitor.php b/src/JMS/Serializer/JsonSerializationVisitor.php
index 5f5caa7..5ef0d7e 100644
--- a/src/JMS/Serializer/JsonSerializationVisitor.php
+++ b/src/JMS/Serializer/JsonSerializationVisitor.php
@@ -65,7 +65,7 @@ class JsonSerializationVisitor extends GenericSerializationVisitor
     public function endVisitingObject(ClassMetadata $metadata, $data, array $type, Context $context)
     {
         $rs = parent::endVisitingObject($metadata, $data, $type, $context);
-
+        /**
         // Force JSON output to "{}" instead of "[]" if it contains either no properties or all properties are null.
         if (empty($rs)) {
             $rs = new \ArrayObject();
@@ -74,7 +74,7 @@ class JsonSerializationVisitor extends GenericSerializationVisitor
                 $this->setRoot(clone $rs);
             }
         }
-
+        //*/
         return $rs;
     }
 }
